File: mvp_js/crypto-roundrunner/src/rng.js
// rng.js
//
// Provably-fair commit-reveal outcome generation. The casino is a crypto
// product, so this isn't just an internal correctness tool — it's also a
// USER-FACING TRUST FEATURE. Players verify rounds independently. Treat it
// like a contract: the commitment is published before the round, the seed is
// revealed afterwards, and the same inputs always produce the same roll.

import { randomBytes, createHmac, createHash } from "node:crypto";

// Outcome space: roll in [0, ROLL_SPACE). 10000 == basis-point granularity.
export const ROLL_SPACE = 10_000;

// Generate a fresh server seed and publish its commitment. Use the OS CSPRNG
// (`crypto.randomBytes`) — never `Math.random`, which is not cryptographically
// secure and is fully predictable from a few samples.
export function newServerSeed() {
  const seed = randomBytes(32);
  return { seed, commitment: commit(seed) };
}

export function commit(seed) {
  return createHash("sha256").update(seed).digest("hex");
}

// Deterministic roll from (serverSeed, clientSeed, nonce). Same inputs ALWAYS
// produce the same output — that's what makes a round auditable.
//
// nonce is a BigInt because round counts can exceed Number.MAX_SAFE_INTEGER
// over the long life of a platform; we want one type, not two.
export function generate(serverSeed, clientSeed, nonce) {
  if (typeof nonce !== "bigint") {
    throw new TypeError(`nonce must be BigInt, got ${typeof nonce}`);
  }
  const mac = createHmac("sha256", serverSeed);
  mac.update(`${clientSeed}:${nonce.toString()}`);
  const digest = mac.digest();
  // First 4 bytes -> uint32 -> [0, ROLL_SPACE). Modulo bias is ~2 × 10^-6 here
  // (ROLL_SPACE / 2^32), negligible; rejection sampling is the strict version.
  const n = digest.readUInt32BE(0);
  return n % ROLL_SPACE;
}

// Verify a previously-played round given the REVEALED server seed. Both the
// commitment check and the roll recomputation must pass. A player or auditor
// runs this; the server doesn't have to be trusted to do it for them.
export function verify({ commitment, serverSeed, clientSeed, nonce, claimedRoll }) {
  if (commit(serverSeed) !== commitment) return false;
  return generate(serverSeed, clientSeed, nonce) === claimedRoll;
}

File: mvp_go/roundrunner/internal/rng/rng.go
// Package rng implements provably-fair outcome generation using a
// commit-reveal scheme.
//
// The flow, and why each step exists:
//
//  1. Before the round, the server generates a random serverSeed and publishes
//     Commitment = SHA256(serverSeed). The hash reveals nothing about the seed
//     but BINDS the server to it: the server can no longer choose the seed
//     after seeing the player's input.
//  2. The player supplies a clientSeed. Because the server can't predict it,
//     the server can't have pre-picked a favourable serverSeed for this round.
//  3. Outcome = HMAC-SHA256(serverSeed, clientSeed || ":" || nonce). This is
//     DETERMINISTIC: identical inputs always produce the same roll. Determinism
//     is what makes a round auditable and replayable.
//  4. After the round the server REVEALS serverSeed. Anyone can check that
//     SHA256(serverSeed) == Commitment and recompute the roll. If both hold,
//     the result was fixed before the player committed and was not tampered
//     with afterwards. That is "provably fair".
package rng

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"fmt"
)

// RollSpace is the size of the outcome space. A Roll is a uniform integer in
// [0, RollSpace), i.e. basis points of the unit interval.
const RollSpace = 10000

// Roll is a single generated outcome in [0, RollSpace).
type Roll uint32

// NewServerSeed returns a fresh 32-byte server seed and its commitment.
func NewServerSeed() (seed []byte, commitment string, err error) {
	seed = make([]byte, 32)
	if _, err = rand.Read(seed); err != nil {
		return nil, "", fmt.Errorf("rng: read entropy: %w", err)
	}
	return seed, Commit(seed), nil
}

// Commit returns SHA256(seed) as hex - the value published before the round.
func Commit(seed []byte) string {
	sum := sha256.Sum256(seed)
	return hex.EncodeToString(sum[:])
}

// Generate derives the deterministic roll for a seed pair and nonce.
//
// The nonce is a uint64. The JS sibling uses a BigInt here because its Number
// type loses integer precision past 2^53; Go's uint64 holds ~1.8 × 10^19
// rounds without loss, which outlives any real platform, so no big.Int is
// needed for the counter.
func Generate(serverSeed []byte, clientSeed string, nonce uint64) Roll {
	mac := hmac.New(sha256.New, serverSeed)
	fmt.Fprintf(mac, "%s:%d", clientSeed, nonce)
	digest := mac.Sum(nil)

	// Map the first 4 bytes to a uint32, then into [0, RollSpace). The modulo
	// introduces a negligible bias (RollSpace / 2^32 ~= 2e-6); a system that
	// truly cares would use rejection sampling. Kept simple here on purpose.
	n := binary.BigEndian.Uint32(digest[:4])
	return Roll(n % RollSpace)
}

// Verify re-runs the commitment check and roll derivation. A player or auditor
// calls this with the revealed seed to confirm the round was fair.
func Verify(commitment string, serverSeed []byte, clientSeed string, nonce uint64, claimed Roll) bool {
	if Commit(serverSeed) != commitment {
		return false
	}
	return Generate(serverSeed, clientSeed, nonce) == claimed
}
